third_party/libunwindstack/tests/PeCoffUnwindInfoUnwinderX86_64Test.cpp

```diff
@@ -0,0 +1,371 @@
+/*
+ * Copyright (C) 2022 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "PeCoffUnwindInfoUnwinderX86_64.h"
+
+#include <algorithm>
+#include <limits>
+
+#include "unwindstack/Error.h"
+#include "unwindstack/MachineX86_64.h"
+#include "unwindstack/RegsX86_64.h"
+#include "utils/MemoryFake.h"
+
+#include "PeCoffEpilog.h"
+#include "PeCoffRuntimeFunctions.h"
+#include "PeCoffUnwindInfoEvaluator.h"
+#include "PeCoffUnwindInfos.h"
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+namespace unwindstack {
+
+class MockPeCoffRuntimeFunctions : public PeCoffRuntimeFunctions {
+ public:
+  MockPeCoffRuntimeFunctions() = default;
+
+  MOCK_METHOD(bool, Init, (uint64_t, uint64_t), (override));
+  MOCK_METHOD(bool, FindRuntimeFunction, (uint64_t, RuntimeFunction*), (const, override));
+};
+
+class MockPeCoffUnwindInfos : public PeCoffUnwindInfos {
+ public:
+  MockPeCoffUnwindInfos() = default;
+
+  MOCK_METHOD(bool, GetUnwindInfo, (uint64_t, UnwindInfo*), (override));
+};
+
+class MockPeCoffEpilog : public PeCoffEpilog {
+ public:
+  MockPeCoffEpilog() = default;
+
+  MOCK_METHOD(bool, Init, (), (override));
+  MOCK_METHOD(bool, DetectAndHandleEpilog, (uint64_t, uint64_t, uint64_t, Memory*, Regs*),
+              (override));
+
+  void FailWithError(ErrorCode error_code) { last_error_.code = error_code; }
+};
+
+class MockPeCoffUnwindInfoEvalutor : public PeCoffUnwindInfoEvaluator {
+ public:
+  MockPeCoffUnwindInfoEvalutor() = default;
+
+  MOCK_METHOD(bool, Eval, (Memory*, Regs*, const UnwindInfo&, PeCoffUnwindInfos*, uint64_t),
+              (override));
+};
+
+class PeCoffUnwindInfoUnwinderX86_64Fake : public PeCoffUnwindInfoUnwinderX86_64 {
+ public:
+  PeCoffUnwindInfoUnwinderX86_64Fake()
+      : PeCoffUnwindInfoUnwinderX86_64(nullptr, 0, 0, 0, 0, 0, {}) {}
+
+  void SetFakeRuntimeFuntions(PeCoffRuntimeFunctions* runtime_functions) {
+    this->runtime_functions_.reset(runtime_functions);
+  }
+
+  void SetFakeUnwindInfos(PeCoffUnwindInfos* unwind_infos) {
+    this->unwind_infos_.reset(unwind_infos);
+  }
+
+  void SetFakeEpilog(PeCoffEpilog* epilog) { this->epilog_.reset(epilog); }
+
+  void SetFakeUnwindInfoEvaluator(PeCoffUnwindInfoEvaluator* unwind_info_evaluator) {
+    this->unwind_info_evaluator_.reset(unwind_info_evaluator);
+  }
```

Comment:
I think these should accept `unique_ptr`s and move them to the member variables. These allows you to use `make_unique` instead of `new` in the tests below, which we want to avoid as much as possible (indeed, using `new` relies on knowing that the class will manage the lifetime of these pointers).

Reply:
Done, excellent suggestion.

Reply:
Please take a look if this looks alright now.